servicemanager/servicemanager.go

```
@@ -0,0 +1,260 @@
+
+	if urlVal.Scheme != "file" {
+		if serviceImage, err = image.Download(context.Background(), sm.downloadDir, imageURL); err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		defer os.RemoveAll(serviceImage)
+	} else {
+		serviceImage = urlVal.Path
+	}
+
+	if err = image.CheckFileInfo(context.Background(), serviceImage, fileInfo); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	// create install dir
+	log.WithFields(log.Fields{"serviceID": newService.ServiceID}).Debug("Install service")
+
+	newService.ImagePath, err = ioutil.TempDir(sm.servicesDir, "")
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	defer func() {
+		if err != nil {
+			_ = os.RemoveAll(newService.ImagePath)
+		}
+	}()
+
+	if err = imageutils.UnpackTarImage(serviceImage, newService.ImagePath); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = validateUnpackedImage(newService.ImagePath); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if isServiceExist {
+		newService.GID = serviceFromStorage.GID
+	} else {
+		newService.GID, err = sm.gidPool.GetFreeID()
+		if err != nil {
+			return aoserrors.Wrap(err)
+		}
+	}
+
+	if err = sm.prepareServiceFS(newService.ImagePath, newService.GID); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	newService.ManifestDigest, err = getManifestChecksum(newService.ImagePath)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = sm.serviceInfoProvider.AddService(newService); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (sm *ServiceManager) prepareServiceFS(imagePath string, gid int) (err error) {
+	imageParts, err := getImageParts(imagePath)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	tmpTar := path.Join(imagePath, "tmpTar")
+	if err = os.Rename(imageParts.ServiceFSPath, tmpTar); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	// unpack rootfs layer
+	if err = imageutils.UnpackTarImage(tmpTar, imageParts.ServiceFSPath); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	os.RemoveAll(tmpTar)
```

Error is not checked.

---

added